### PR TITLE
[RLlib; Offline RL] RLUnplugged example on new API stack.

### PR DESCRIPTION
**Pull request by simonsays1980**
## Why are these changes needed?

The new `Offline RL API`  uses native `ray.data.Dataset`s to scale to massive datasets with different formats and encodings. To ensure that learning indeed scales to massive data and the API can be used with the most notable Offline RL datasets this PR applies the new `Offline RL API` to the `RLUnplugged`  data. In the tuned example data from Atari `Pong` is used with a single `TfRecord`s file from the GCS bucket.

The data is downloaded automatically into a temporary folder. To decode the PNG images this example defines a `ConnectorV2` that decodes the images in the learner connector pipeline and stacks frames.

This example show specifically how to scale out on massive data and use the `AlgorithmConfig.offline_data`'s arguments to fine-tune scaling with `ray.data`. See for more infos `ray.data`. 
 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/tuned_examples/bc/benchmark_atari_pong_bc.py [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+            "vf_share_layers": True,
+            "conv_filters": [[16, 4, 2], [32, 4, 2], [64, 4, 2], [128, 4, 2]],
+            "conv_activation": "relu",
+            "post_fcnet_hiddens": [256],
+            "uses_new_env_runners": True,
+        }
+    )
+)
+
+# TODO (simon): Change to use the `run_rllib_example` function as soon as tuned.
```
Comment (sven1977) [inline rllib/tuned_examples/bc/benchmark_atari_pong_bc.py:251]: nit: TODO? Or still WIP?

File: rllib/tuned_examples/bc/benchmark_atari_pong_bc.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+game = "Pong"
+# There are many run numbers, we choose the first one for demonstration. This
+# can be chosen by users. To use all data use a list of file paths (see
+# `num_shards`) and its usage further below.
+run_number = 1
+# num_shards = 1
+
+# Make the temporary directory for the downloaded data.
+tmp_path = "/tmp/atari"
+Path(tmp_path).joinpath(game).mkdir(exist_ok=True, parents=True)
```
Comment (sven1977) [inline rllib/tuned_examples/bc/benchmark_atari_pong_bc.py:153]: super nit: should we use the "/" op of pathlib.Path here?

```
tmp_path = Path("/tmp/atari") / game
tmp_path.mkdir(...)
destination_file_name = tmp_path / "run_...."
```

File: rllib/tuned_examples/bc/benchmark_atari_pong_bc.py [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+args = parser.parse_args()
+
+# We only use the Atari game `Pong` here. Users can choose other Atari
+# games and set here the name. This will download `TfRecords` dataset from GCS.
+game = "Pong"
+# There are many run numbers, we choose the first one for demonstration. This
+# can be chosen by users. To use all data use a list of file paths (see
+# `num_shards`) and its usage further below.
+run_number = 1
+# num_shards = 1
```
Comment (sven1977) [inline rllib/tuned_examples/bc/benchmark_atari_pong_bc.py:149]: Should the be commented out? `# num_shards = 1`

File: rllib/tuned_examples/bc/benchmark_atari_pong_bc.py [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+    return DecodeObservations()
+
+
+# Wrap the environment used in evalaution into `RLlib`'s Atari Wrapper
+# that automatically stacks frames and converts to the dimension used
+# in the collection of the `rl_unplugged` data.
+def _env_creator(cfg):
+    return wrap_atari_for_new_api_stack(
+        gym.make("ALE/Pong-v5", **cfg),
+        # Perform frame-stacking through ConnectorV2 API.
```
Comment (sven1977) [inline rllib/tuned_examples/bc/benchmark_atari_pong_bc.py:128]: Wait, if we say: framestack=4 here then we do NOT perform frame-stacking through ConnectorV2 here.

-> Change comment by removing the ConnectorV2 statement.

Note: At this point, the gain in performance when using the connector is minimal, so can probably be neglected for simplicity (easier for user to do framestacking in env-wrapper).

File: rllib/tuned_examples/bc/benchmark_atari_pong_bc.py [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+from ray.rllib.env.wrappers.atari_wrappers import wrap_atari_for_new_api_stack
+from ray.rllib.utils.annotations import override
+from ray.rllib.utils.test_utils import (
+    add_rllib_example_script_args,
+)
+from ray import tune
+
+
+class DecodeObservations(ConnectorV2):
+    def __init__(
```
Comment (sven1977) [inline rllib/tuned_examples/bc/benchmark_atari_pong_bc.py:36]: Add a (small) docstring here that explains what this connector does.

File: rllib/tuned_examples/bc/benchmark_atari_pong_bc.py [hunk 6]
```diff
@@ -0,0 +1,257 @@
+"""
+schema={
```
Comment (sven1977) [inline rllib/tuned_examples/bc/benchmark_atari_pong_bc.py:2]: Is this needed here? If this is an explanation of the schema, maybe move this down into the config section (where we explain the schema translation)?

File: rllib/algorithms/algorithm_config.py [hunk 7]
```diff
@@ -2499,6 +2499,10 @@ def offline_data(
             self.input_read_method_kwargs = input_read_method_kwargs
         if input_read_schema is not NotProvided:
             self.input_read_schema = input_read_schema
+        if map_batches_kwargs is not NotProvided:
```
Comment (sven1977) [inline rllib/algorithms/algorithm_config.py:2502]: Ah, yes, this was missing :)

**Review by sven1977 — COMMENTED** [review]
Looks great! Just some nits, mostly enhancing/adding comments and explanations.

File: rllib/tuned_examples/bc/benchmark_atari_pong_bc.py [hunk 8]
```diff
@@ -0,0 +1,257 @@
+"""
```
Comment (sven1977) [inline rllib/tuned_examples/bc/benchmark_atari_pong_bc.py:1]: Should we call this file simply: `pong_bc.py`?

File: rllib/tuned_examples/bc/benchmark_atari_pong_bc.py [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+            as_learner_connector: Whether this connector is part of a Learner connector
+                pipeline, as opposed to an env-to-module pipeline.
+        """
+        super().__init__(
+            input_observation_space=input_observation_space,
+            input_action_space=input_action_space,
+            **kwargs,
+        )
+
+        self._multi_agent = multi_agent
```
[the remark this answers is not in the document]
Comment (simonsays1980) [reply rllib/tuned_examples/bc/benchmark_atari_pong_bc.py:63]: No, we don't. It was not clear to me, yet, that this does not need to be set when in single-agent mode.

File: rllib/tuned_examples/bc/benchmark_atari_pong_bc.py [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+                num_items=len(sa_episode),
+                single_agent_episode=sa_episode,
+            )
+
+        return data
+
+
+# Make the learner connector.
+def _make_learner_connector(observation_space, action_space):
+    return DecodeObservations()
```
[the remark this answers is not in the document]
Comment (simonsays1980) [reply rllib/tuned_examples/bc/benchmark_atari_pong_bc.py:119]: I like this more! Thanks!

File: rllib/tuned_examples/bc/benchmark_atari_pong_bc.py (hunk 1 above)
Comment (simonsays1980) [reply rllib/tuned_examples/bc/benchmark_atari_pong_bc.py:251]: WIP :) It is still not running smoothly and for debugging I am not using tune.

File: rllib/tuned_examples/bc/benchmark_atari_pong_bc.py (hunk 2 above)
Comment (simonsays1980) [reply rllib/tuned_examples/bc/benchmark_atari_pong_bc.py:153]: Yeah, let's do it this is better to read.

File: rllib/tuned_examples/bc/benchmark_atari_pong_bc.py (hunk 3 above)
Comment (simonsays1980) [reply rllib/tuned_examples/bc/benchmark_atari_pong_bc.py:149]: This is actually a number that is now hard coded into the path. We have many runs and for each multiple shards in the bucket - each is a file. For the example I use only a single of these files.

File: rllib/tuned_examples/bc/benchmark_atari_pong_bc.py (hunk 4 above)
Comment (simonsays1980) [reply rllib/tuned_examples/bc/benchmark_atari_pong_bc.py:128]: Thanks for the hint. WIll remove this comment.

**Review by sven1977 — APPROVED** [review]
Looks good to me now! Thanks @simonsays1980

**Pull request merged**